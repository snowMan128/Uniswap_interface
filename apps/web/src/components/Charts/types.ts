import { SeriesDataItemTypeMap, SeriesOptionsMap, Time } from 'lightweight-charts'

export type SeriesDataItemType = SeriesDataItemTypeMap<Time>[keyof SeriesOptionsMap]
