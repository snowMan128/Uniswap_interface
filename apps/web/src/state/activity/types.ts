import { SupportedInterfaceChainId } from 'constants/chains'
import { SignatureDetails } from 'state/signatures/types'
import { SerializableTransactionReceipt, TransactionDetails } from '../transactions/types'

interface BaseUpdate<T> {
  type: string
  chainId: SupportedInterfaceChainId
  original: T
  update?: Partial<T>
  receipt?: SerializableTransactionReceipt
}

export interface TransactionUpdate extends BaseUpdate<TransactionDetails> {
  type: 'transaction'
  receipt: SerializableTransactionReceipt
}

interface OrderUpdate extends BaseUpdate<SignatureDetails> {
  type: 'signature'
}

export type ActivityUpdate = TransactionUpdate | OrderUpdate
export type OnActivityUpdate = (update: ActivityUpdate) => void
