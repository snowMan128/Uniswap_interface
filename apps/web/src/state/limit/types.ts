export enum Expiry {
  Day = 'Day',
  Week = 'Week',
  Month = 'Month',
  Year = 'Year',
}

export interface LimitState {
  readonly inputAmount: string
  readonly outputAmount: string
  readonly expiry: Expiry
  readonly limitPrice: string
  readonly limitPriceInverted: boolean

  // The form should autofill in the market price between two currencies unless the user has
  // already manually edited the price for that currency pair
  readonly limitPriceEdited: boolean

  // The limit form has 3 fields, but only two of them can be independent at a time.
  // Always prefer `marketPrice` be independent, so either derive the input amount or the output amount
  readonly isInputAmountFixed: boolean
}
