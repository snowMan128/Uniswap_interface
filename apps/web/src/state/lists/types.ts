import { TokenList } from '@uniswap/token-lists'

export interface ListsState {
  readonly byUrl: {
    readonly [url: string]: {
      readonly current: TokenList | null
      readonly pendingUpdate: TokenList | null
      readonly loadingRequestId: string | null
      readonly error: string | null
    }
  }
  // this contains the default list of lists from the last time the updateVersion was called, i.e. the app was reloaded
  readonly lastInitializedDefaultListOfLists?: string[]
}
