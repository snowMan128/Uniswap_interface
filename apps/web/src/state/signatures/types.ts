import { UniswapXOrderStatus } from 'lib/hooks/orders/types'

import { ExactInputSwapTransactionInfo, ExactOutputSwapTransactionInfo } from '../transactions/types'

export enum SignatureType {
  SIGN_UNISWAPX_ORDER = 'signUniswapXOrder',
  SIGN_LIMIT = 'signLimit',
}

interface BaseSignatureFields {
  type: SignatureType
  id: string
  addedTime: number
  chainId: number
  expiry?: number
  offerer: string
}

export interface UniswapXOrderDetails extends BaseSignatureFields {
  type: SignatureType
  orderHash: string
  status: UniswapXOrderStatus
  swapInfo: (ExactInputSwapTransactionInfo | ExactOutputSwapTransactionInfo) & { isUniswapXOrder: true }
  txHash?: string
  encodedOrder?: string
}

export type SignatureDetails = UniswapXOrderDetails
