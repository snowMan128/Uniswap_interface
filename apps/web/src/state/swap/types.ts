import { ChainId, Currency, CurrencyAmount, Percent } from '@uniswap/sdk-core'
import { Field, SwapTab } from 'components/swap/constants'
import { parsedQueryString } from 'hooks/useParsedQueryString'
import { ParsedQs } from 'qs'
import { Dispatch, ReactNode, SetStateAction, createContext } from 'react'
import { InterfaceTrade, TradeState } from 'state/routing/types'

export type SwapInfo = {
  currencies: { [field in Field]?: Currency }
  currencyBalances: { [field in Field]?: CurrencyAmount<Currency> }
  inputTax: Percent
  outputTax: Percent
  outputFeeFiatValue?: number
  parsedAmount?: CurrencyAmount<Currency>
  inputError?: ReactNode
  trade: {
    trade?: InterfaceTrade
    state: TradeState
    uniswapXGasUseEstimateUSD?: number
    error?: any
    swapQuoteLatency?: number
  }
  allowedSlippage: Percent
  autoSlippage: Percent
}

type SwapContextType = {
  swapState: SwapState
  derivedSwapInfo: SwapInfo
  setSwapState: Dispatch<SetStateAction<SwapState>>
}

function parseTokenAmountURLParameter(urlParam: any): string {
  return typeof urlParam === 'string' && !isNaN(parseFloat(urlParam)) ? urlParam : ''
}

export function parseIndependentFieldURLParameter(urlParam: any): Field {
  return typeof urlParam === 'string' && urlParam.toLowerCase() === 'output' ? Field.OUTPUT : Field.INPUT
}

export function queryParametersToSwapState(parsedQs: ParsedQs): SwapState {
  const typedValue = parseTokenAmountURLParameter(parsedQs.exactAmount)
  const independentField = parseIndependentFieldURLParameter(parsedQs.exactField)

  return {
    typedValue,
    independentField,
  }
}

export const EMPTY_DERIVED_SWAP_INFO: SwapInfo = Object.freeze({
  currencies: {},
  currencyBalances: {},
  inputTax: new Percent(0),
  outputTax: new Percent(0),
  autoSlippage: new Percent(0),
  allowedSlippage: new Percent(0),
  trade: {
    state: TradeState.LOADING,
  },
})

export const initialSwapState: SwapState = queryParametersToSwapState(parsedQueryString())

export const SwapContext = createContext<SwapContextType>({
  swapState: initialSwapState,
  derivedSwapInfo: EMPTY_DERIVED_SWAP_INFO,
  setSwapState: () => undefined,
})

type SwapAndLimitContextType = {
  currencyState: CurrencyState
  prefilledState: {
    inputCurrency?: Currency
    outputCurrency?: Currency
  }
  setCurrencyState: Dispatch<SetStateAction<CurrencyState>>
  currentTab: SwapTab
  setCurrentTab: Dispatch<SetStateAction<SwapTab>>
  // The chainId of the page/context - can be different from the connected Chain ID if the
  // page is displaying content for a different chain
  chainId?: ChainId
}

export const SwapAndLimitContext = createContext<SwapAndLimitContextType>({
  currencyState: {
    inputCurrency: undefined,
    outputCurrency: undefined,
  },
  setCurrencyState: () => undefined,
  prefilledState: {
    inputCurrency: undefined,
    outputCurrency: undefined,
  },
  chainId: ChainId.MAINNET,
  currentTab: SwapTab.Swap,
  setCurrentTab: () => undefined,
})

export interface SerializedCurrencyState {
  inputCurrencyId?: string
  outputCurrencyId?: string
}

// shared state between Swap and Limit
export interface CurrencyState {
  inputCurrency?: Currency
  outputCurrency?: Currency
}

export interface SwapState {
  readonly independentField: Field
  readonly typedValue: string
}
