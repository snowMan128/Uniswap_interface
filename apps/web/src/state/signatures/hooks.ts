import { ChainId } from '@uniswap/sdk-core'
import { useWeb3React } from '@web3-react/core'
import { useCallback, useMemo } from 'react'
import { useDispatch } from 'react-redux'
import { useAppSelector } from 'state/hooks'
import { OFFCHAIN_ORDER_TYPE_TO_SIGNATURE_TYPE, OffchainOrderType } from 'state/routing/types'
import { UniswapXOrderStatus } from 'types/uniswapx'
import { addSignature } from './reducer'
import { SignatureDetails, SignatureType, UniswapXOrderDetails } from './types'

export function useAllSignatures(): { [id: string]: SignatureDetails } {
  const { account } = useWeb3React()
  const signatures = useAppSelector((state) => state.signatures) ?? {}
  if (!account || !signatures[account]) return {}
  return signatures[account]
}

export function usePendingOrders(): UniswapXOrderDetails[] {
  const signatures = useAllSignatures()
  return useMemo(() => {
    return Object.values(signatures).filter(isPendingOrder)
  }, [signatures])
}

export function useOrder(orderHash: string): UniswapXOrderDetails | undefined {
  const signatures = useAllSignatures()
  return useMemo(() => {
    const order = signatures[orderHash]
    if (
      !order ||
      ![SignatureType.SIGN_UNISWAPX_ORDER, SignatureType.SIGN_UNISWAPX_V2_ORDER, SignatureType.SIGN_LIMIT].includes(
        order.type as SignatureType
      )
    )
      return undefined
    return order
  }, [signatures, orderHash])
}

export function useAddOrder() {
  const dispatch = useDispatch()

  return useCallback(
    (
      offerer: string,
      orderHash: string,
      chainId: ChainId,
      expiry: number,
      swapInfo: UniswapXOrderDetails['swapInfo'],
      encodedOrder: string,
      offchainOrderType: OffchainOrderType
    ) => {
      dispatch(
        addSignature({
          type: OFFCHAIN_ORDER_TYPE_TO_SIGNATURE_TYPE[offchainOrderType],
          offerer,
          id: orderHash,
          chainId,
          expiry,
          orderHash,
          swapInfo,
          status: UniswapXOrderStatus.OPEN,
          addedTime: Date.now(),
          encodedOrder,
        })
      )
    },
    [dispatch]
  )
}

export function isFinalizedOrder(orderStatus: UniswapXOrderStatus) {
  return orderStatus !== UniswapXOrderStatus.OPEN
}

export function isOnChainOrder(orderStatus: UniswapXOrderStatus) {
  return orderStatus === UniswapXOrderStatus.FILLED
}

function isPendingOrder(signature: SignatureDetails): signature is UniswapXOrderDetails {
  return (
    (signature.type === SignatureType.SIGN_UNISWAPX_ORDER ||
      signature.type === SignatureType.SIGN_UNISWAPX_V2_ORDER ||
      signature.type === SignatureType.SIGN_LIMIT) &&
    signature.status === UniswapXOrderStatus.OPEN
  )
}
