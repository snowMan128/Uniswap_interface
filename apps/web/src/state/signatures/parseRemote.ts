import { TradeType } from '@uniswap/sdk-core'
import { asSupportedChain } from 'constants/chains'
import { parseUnits } from 'ethers/lib/utils'
import { gqlToCurrency, logSentryErrorForUnsupportedChain, supportedChainIdFromGQLChain } from 'graphql/data/util'
import store from 'state'
import { TransactionType as LocalTransactionType } from 'state/transactions/types'
import { UniswapXOrderStatus } from 'types/uniswapx'
import {
  AssetActivityPartsFragment,
  SwapOrderDetailsPartsFragment,
  SwapOrderStatus,
  SwapOrderType,
} from 'uniswap/src/data/graphql/uniswap-data-api/__generated__/types-and-hooks'
import { currencyId } from 'utils/currencyId'
import { addSignature } from './reducer'
import { SignatureDetails, SignatureType, UniswapXOrderDetails } from './types'

export type OrderActivity = AssetActivityPartsFragment & { details: SwapOrderDetailsPartsFragment }

const SIGNATURE_TYPE_MAP: { [key in SwapOrderType]: SignatureType } = {
  [SwapOrderType.Limit]: SignatureType.SIGN_LIMIT,
  [SwapOrderType.Dutch]: SignatureType.SIGN_UNISWAPX_ORDER,
  [SwapOrderType.DutchV2]: SignatureType.SIGN_UNISWAPX_V2_ORDER,
}

const ORDER_STATUS_MAP: { [key in SwapOrderStatus]: UniswapXOrderStatus } = {
  [SwapOrderStatus.Open]: UniswapXOrderStatus.OPEN,
  [SwapOrderStatus.Expired]: UniswapXOrderStatus.EXPIRED,
  [SwapOrderStatus.Error]: UniswapXOrderStatus.ERROR,
  [SwapOrderStatus.InsufficientFunds]: UniswapXOrderStatus.INSUFFICIENT_FUNDS,
  [SwapOrderStatus.Filled]: UniswapXOrderStatus.FILLED,
  [SwapOrderStatus.Cancelled]: UniswapXOrderStatus.CANCELLED,
}

export function parseRemote({ chain, details, timestamp }: OrderActivity): SignatureDetails {
  const supportedChain = asSupportedChain(supportedChainIdFromGQLChain(chain))
  if (!supportedChain) {
    const error = new Error('Invalid activity from unsupported chain received from GQL')
    logSentryErrorForUnsupportedChain({ extras: { details }, errorMessage: error.message })
    throw error
  }

  const status = ORDER_STATUS_MAP[details.orderStatus]
  const isFilled = status == UniswapXOrderStatus.FILLED

  const inputTokenQuantity = parseUnits(details.inputTokenQuantity, details.inputToken.decimals).toString()
  const outputTokenQuantity = parseUnits(details.outputTokenQuantity, details.outputToken.decimals).toString()

  if (inputTokenQuantity === '0' || outputTokenQuantity === '0') {
    // TODO(WEB-3765): This is a temporary mitigation for a bug where the backend sends "0.000000" for small amounts.
    throw new Error('Invalid activity received from GQL')
  }

  const signature: UniswapXOrderDetails = {
    id: details.id,
    type: SIGNATURE_TYPE_MAP[details.swapOrderType],
    offerer: details.offerer,
    chainId: supportedChain,
    orderHash: details.hash,
    expiry: details.expiry,
    encodedOrder: details.encodedOrder,
    status,
    addedTime: timestamp,
    // only if completed
    txHash: isFilled ? details.hash : undefined,

    swapInfo: {
      isUniswapXOrder: true,
      type: LocalTransactionType.SWAP,
      // This doesn't affect the display, but we don't know this value from the remote activity.
      tradeType: TradeType.EXACT_INPUT,
      inputCurrencyId: currencyId(gqlToCurrency(details.inputToken)),
      outputCurrencyId: currencyId(gqlToCurrency(details.outputToken)),
      inputCurrencyAmountRaw: inputTokenQuantity,
      expectedOutputCurrencyAmountRaw: outputTokenQuantity,
      minimumOutputCurrencyAmountRaw: outputTokenQuantity,
      settledOutputCurrencyAmountRaw: isFilled ? outputTokenQuantity : undefined,
    },
  }

  if (status === UniswapXOrderStatus.OPEN) {
    store.dispatch(addSignature(signature))
  }

  return signature
}
