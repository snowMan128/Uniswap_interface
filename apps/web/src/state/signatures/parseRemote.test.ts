import { OrderActivity } from 'state/signatures/types'
import { MockExpiredUniswapXOrder, MockFilledUniswapXOrder, MockOpenUniswapXOrder } from './fixtures'
import { parseRemote } from './parseRemote'

describe('parseRemote', () => {
  it('should parse open UniswapX order', () => {
    const result = parseRemote(MockOpenUniswapXOrder as OrderActivity)
    expect(result).toMatchSnapshot()
  })
  it('should parse expired UniswapX order', () => {
    const result = parseRemote(MockExpiredUniswapXOrder as OrderActivity)
    expect(result).toMatchSnapshot()
  })
  it('should parse filledUniswapX order', () => {
    const result = parseRemote(MockFilledUniswapXOrder as OrderActivity)
    expect(result).toMatchSnapshot()
  })
})
