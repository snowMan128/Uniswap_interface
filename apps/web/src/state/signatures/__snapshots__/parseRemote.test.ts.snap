// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRemote should parse expired UniswapX order 1`] = `
Object {
  "addedTime": 10000,
  "chainId": 1,
  "encodedOrder": undefined,
  "expiry": undefined,
  "id": "someId",
  "offerer": "someOfferer",
  "orderHash": "someHash",
  "status": "expired",
  "swapInfo": Object {
    "expectedOutputCurrencyAmountRaw": "200000000000000000000",
    "inputCurrencyAmountRaw": "100000000000000000000",
    "inputCurrencyId": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
    "isUniswapXOrder": true,
    "minimumOutputCurrencyAmountRaw": "200000000000000000000",
    "outputCurrencyId": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    "settledOutputCurrencyAmountRaw": undefined,
    "tradeType": 0,
    "type": 1,
  },
  "txHash": undefined,
  "type": undefined,
}
`;

exports[`parseRemote should parse filledUniswapX order 1`] = `
Object {
  "addedTime": 10000,
  "chainId": 1,
  "encodedOrder": undefined,
  "expiry": undefined,
  "id": "someId",
  "offerer": "someOfferer",
  "orderHash": "someHash",
  "status": "filled",
  "swapInfo": Object {
    "expectedOutputCurrencyAmountRaw": "200000000000000000000",
    "inputCurrencyAmountRaw": "100000000000000000000",
    "inputCurrencyId": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
    "isUniswapXOrder": true,
    "minimumOutputCurrencyAmountRaw": "200000000000000000000",
    "outputCurrencyId": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    "settledOutputCurrencyAmountRaw": "200000000000000000000",
    "tradeType": 0,
    "type": 1,
  },
  "txHash": "someHash",
  "type": undefined,
}
`;

exports[`parseRemote should parse open UniswapX order 1`] = `
Object {
  "addedTime": 10000,
  "chainId": 1,
  "encodedOrder": undefined,
  "expiry": undefined,
  "id": "someId",
  "offerer": "someOfferer",
  "orderHash": "someHash",
  "status": "open",
  "swapInfo": Object {
    "expectedOutputCurrencyAmountRaw": "200000000000000000000",
    "inputCurrencyAmountRaw": "100000000000000000000",
    "inputCurrencyId": "0x6B175474E89094C44Da98b954EedeAC495271d0F",
    "isUniswapXOrder": true,
    "minimumOutputCurrencyAmountRaw": "200000000000000000000",
    "outputCurrencyId": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    "settledOutputCurrencyAmountRaw": undefined,
    "tradeType": 0,
    "type": 1,
  },
  "txHash": undefined,
  "type": undefined,
}
`;
